### ISSUE TITLE
[Question] Does AsyncWorker create multiple threads?

### ISSUE THREAD

**Issue by mithusingh32** [nodejs/node-addon-api#1084]
I've been using AsynWorker to handle my processing, but there as an issue that it initialized a third-party library and I don't want that to happen. So I am rewriting my addon to use Napi::Addon and initialize the 3rd party library once. And then call the function below. 
```c++
Napi::Value MyAddon::getAudioData(const Napi::CallbackInfo &info)
    {
        Napi::Env env = info.Env();
        if (info.Length() < 2)
        {
            Napi::TypeError::New(env, "Wrong number of arguments")
                .ThrowAsJavaScriptException();
            return env.Null();
        }

        if (!info[0].IsString() || !info[1].IsFunction())
        {
            Napi::TypeError::New(env, "Wrong arguments").ThrowAsJavaScriptException();
            return env.Null();
        }

        Napi::Function callback = info[1].As<Napi::Function>();
        MyAnalyzer::MyAnalyzer *Analyzer = new MyAnalyzer::MyAnalyzer(callback, info[0].As<Napi::String>());
        AudioAnalyzer->Queue(); 
        return Napi::String::New(info.Env(), "MyData");
    }
```
The final return is an object with the data I need. 

The question is lets say I want 5 files to be processed. Will this code create 5 threads and execute the processing or will it `Queue` up 5 files and process them correctly on a single thread?

**Comment by mhdawson** [comment]
Queue in AsyncWorker calls napi_queue_async_work
https://github.com/nodejs/node-addon-api/blob/cb228418e6df1d613bc262504b856f95c69900e1/napi-inl.h#L4756

that in turn calls the ScheduleWork on the uvimpl::Work*

https://github.com/nodejs/node/blob/e58592e49fe1374e46e3d4825198e51784dcde2d/src/node_api.cc#L1170

where Work is defined at https://github.com/nodejs/node/blob/e58592e49fe1374e46e3d4825198e51784dcde2d/src/node_api.cc#L1038

Looking at ThreadPoolWork I believe the only queuing that will take place is if there are already more tasks queued than the value set for the UV_THREADPOOL_SIZE environment variable (default is 4).

I'd have to test out the code to be sure, but I think the answer is no, they will run up to 4 in parallel (or more if you have changed UV_THREADPOOL_SIZE). If you want them to be queued one after the other you will need to manage a queue that kicks one off after the last one has completed.

@gabrielschulhof, @KevinEady  as a sanity check.

**Comment by mithusingh32** [comment]
Thanks for replying quickly!!
That is what I thought. But every time I read the docs, the more I felt I didn't understand completely.

**Comment by mhdawson** [comment]
@mithusingh32 I agree it's not 100% clear from the docs, we'll have to think about fixing that.
